File: accounting-service/src/main/java/com/acme/accountingsrv/plan/service/PlanService.java
package com.acme.accountingsrv.plan.service;

import com.acme.accountingsrv.plan.PlanStatus;
import com.acme.accountingsrv.plan.dto.PlanWithCountriesDto;
import com.acme.accountingsrv.plan.dto.PlanWithCountDto;
import com.acme.accountingsrv.plan.dto.SavePlanDto;
import com.acme.accountingsrv.plan.dto.PlanDto;
import com.acme.accountingsrv.plan.dto.PlanFilter;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.validation.annotation.Validated;
import reactor.core.publisher.Mono;

import javax.validation.Valid;
import javax.validation.constraints.NotNull;
import java.util.List;
import java.util.UUID;

@Validated
public interface PlanService {
    @PreAuthorize("hasAnyAuthority('ADMIN','ACCOUNTANT')")
    Mono<UUID> create(@Valid SavePlanDto createDto);

    @PreAuthorize("hasAnyAuthority('ADMIN','ACCOUNTANT')")
    Mono<Void> update(UUID id, @Valid SavePlanDto createDto);

    @PreAuthorize("hasAnyAuthority('ADMIN','ACCOUNTANT')")
    Mono<Void> changeStatus(UUID id, @NotNull PlanStatus newStatus);

    @PreAuthorize("isAuthenticated()")
    Mono<PlanWithCountriesDto> findById(UUID id);

    @PreAuthorize("hasAnyAuthority('ADMIN','ACCOUNTANT')")
    Mono<Page<PlanWithCountDto>> find(PlanFilter filter, Pageable pageable);

    Mono<List<PlanDto>> findActive(String country);

}

File: admin-ui/src/plan/planService.js
import restApi, { buildGetFilterParams } from '../common/restApi';

const BASE_URL = '/accounting-service/plans';

const planService = {
    create: function (request) {
        return restApi.post(BASE_URL, request);
    },
    find: function (filter, pageable, sort) {
        return restApi.get(BASE_URL, {
            params: buildGetFilterParams(filter, pageable, sort)
        });
    },
    changeStatus: function (id, request) {
        return restApi.put(`${BASE_URL}/${id}/status`, request);
    },
    findById: function (id) {
        return restApi.get(`${BASE_URL}/${id}`);
    },
    update: function (id, request) {
        return restApi.put(`${BASE_URL}/${id}`, request);
    },
    findActive: function (country) {
        return restApi.get(`${BASE_URL}/active`, {
            params: { country: country }
        });
    },
    findStatistics: function (id) {
        return restApi.get(`${BASE_URL}/${id}/statistics`);
    },
}

export default planService;